contracts/src/CrossChainApplications/NativeTokenBridge/ERC20TokenSource.sol

```diff
@@ -0,0 +1,174 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@subnet-evm-contracts/interfaces/IWarpMessenger.sol";
+import "./IERC20TokenSource.sol";
+import "../../Teleporter/ITeleporterMessenger.sol";
+import "../../Teleporter/ITeleporterReceiver.sol";
+import "../../Teleporter/SafeERC20TransferFrom.sol";
+
+contract ERC20TokenSource is
+    ITeleporterReceiver,
+    IERC20TokenSource,
+    ReentrancyGuard
+{
+    using SafeERC20 for IERC20;
+
+    uint256 public constant MINT_NATIVE_TOKENS_REQUIRED_GAS = 150_000; // TODO this is a placeholder
+    bytes32 public immutable currentBlockchainID;
+    bytes32 public immutable destinationBlockchainID;
+    address public immutable nativeTokenDestinationAddress;
+    address public immutable erc20ContractAddress;
+
+    // Used for sending an receiving Teleporter messages.
+    ITeleporterMessenger public immutable teleporterMessenger;
+
+    constructor(
+        address teleporterMessengerAddress,
+        bytes32 destinationBlockchainID_,
+        address nativeTokenDestinationAddress_,
+        address erc20ContractAddress_
+    ) {
+        currentBlockchainID = WarpMessenger(
+            0x0200000000000000000000000000000000000005
+        ).getBlockchainID();
+
+        require(
+            teleporterMessengerAddress != address(0),
+            "ERC20TokenSource: zero TeleporterMessenger Address"
+        );
+        teleporterMessenger = ITeleporterMessenger(teleporterMessengerAddress);
+
+        require(
+            destinationBlockchainID_ != bytes32(0),
+            "ERC20TokenSource: zero Destination Chain ID"
+        );
+        require(
+            destinationBlockchainID_ != currentBlockchainID,
+            "ERC20TokenSource: Cannot Bridge With Same Blockchain"
+        );
+        destinationBlockchainID = destinationBlockchainID_;
+
+        require(
+            nativeTokenDestinationAddress_ != address(0),
+            "ERC20TokenSource: zero Destination Contract Address"
+        );
+        nativeTokenDestinationAddress = nativeTokenDestinationAddress_;
+
+        require(
+            erc20ContractAddress_ != address(0),
+            "ERC20TokenSource: Invalid ERC20 Contract Address"
+        );
+        erc20ContractAddress = erc20ContractAddress_;
+    }
+
+    /**
+     * @dev See {ITeleporterReceiver-receiveTeleporterMessage}.
+     *
+     * Receives a Teleporter message and routes to the appropriate internal function call.
```

Comment:
This implementation of `receiveTeleporterMessage` only has a single function, so let's document that explicitly in the comment.